Sample Data Index and Plugin Classes

huyaboo opened the pull request:
Signed-off-by: Huy Nguyen <73027756+huyaboo@users.noreply.github.com>

### Description
This PR adds the Index and Plugin classes needed to create/delete indices and plugins.
 
### Issues Resolved
#37 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Flyingliuhub commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+"""
+class SampleDataIndex:
+    def __init__(self, index_name, client, index_body):
+        self.index_name = index_name
+        self.index_body = index_body
+        self.client = client
```
if possible, please add some logic to validate it. Thanks

huyaboo replied:
Done, type checking is enforced for the variables

Flyingliuhub commented on src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py:
```diff
@@ -0,0 +1,145 @@
+            self.payload = dumps(self.payload)
+        
+        # Add authentication headers
+        __auth_plain = environ.get("SAMPLE_DATA_USERNAME") + ":" + environ.get("SAMPLE_DATA_PASSWORD")
```
Could you please provide more detail in the top for SAMPLE_DATA_USERNAME and SAMPLE_DATA_PASSWORD? if possiable to add README.md

huyaboo replied:
I changed logic so now the class uses an authentication class to build the header

Flyingliuhub commented on src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py:
```diff
@@ -0,0 +1,145 @@
+            payload["result_index"] = self.target_index
+        payload = dumps(payload)
+        response = request("POST", endpoint, headers = self.headers, data = payload, verify = False)
+        print(response.text)
```
what's the reposnse.text here? do you have some sample

huyaboo replied:
Yes, I just included samples for each type of response

Flyingliuhub commented on src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py:
```diff
@@ -0,0 +1,145 @@
+        if self.target_index:
+            payload["result_index"] = self.target_index
+        payload = dumps(payload)
+        response = request("POST", endpoint, headers = self.headers, data = payload, verify = False)
```
if request failed, do we catch the exception for the request ?

huyaboo replied:
Done, error catching is impleemented for each request

Flyingliuhub commented on src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py:
```diff
@@ -0,0 +1,145 @@
+
+    # Deletes detector; detector must already exist
+    def delete_detector(self):
+        endpoint = self.url + "detectors/" + self.id
```
if self.id empty, then delete call the decetor ?

huyaboo replied:
I added a try block to ensure that the detector is started before it's stopped

Flyingliuhub commented on src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py:
```diff
@@ -0,0 +1,145 @@
+    - payload: The JSON body of which to call create API requests
+    - base_url: The IP address/url where the OpenSearch backend lies
+"""
+class Plugin:
```
are you planing to add all the plugins here? or create a generic plugin call, then separate file for each plugins ?

CCongWang replied:
+1, you could have a base file for plugin class, and then have separate files for different plugins

huyaboo replied:
Done, all of the classes are on separate files

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+    - client: an OpenSearch Python client object
+"""
+class SampleDataIndex:
+    def __init__(self, index_name, client, index_body):
```
I believe this `__init__` is for class initialization, not for packaging

huyaboo replied:
I do not need an `__init__.py` for this tooling

huyaboo replied:
Done, latest commit defines data type (and wrote validation to enforce type checking)

bandinib-amzn reviewed (changes requested):
I observed for new files you are not adding license header. Is this intended? 

What is plan for adding unit test cases?

CCongWang commented on src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py:
```diff
@@ -0,0 +1,145 @@
+
+    # Utility function to convert payload to a dict
+    def convert_payload(self, payload, plugin_config_directory_name):
+        if type(payload) is str and ".json" not in payload:
```
does it mean that payload is a file but not a json file??

huyaboo replied:
> +1, you could have a base file for plugin class, and then have separate files for different plugins

Done

> does it mean that payload is a file but not a json file??

Comments are updated but this code block sees that this payload is a string (non filename) and will try to convert to a `dict` (because the function assumes the `payload` is a `JSON string`). If this fails, the function knows the input was a non `JSON string`, non `filename string` string.

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+from utils.sample_data_ingestor.sample_data_ingestor import ingest
```
Done, latest commit adheres to the folder structure as seen with the generation tool

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+    def create_index(self):
+        if not self.client.indices.exists(index = self.index_name):
+            self.client.indices.create(index = self.index_name, body=self.index_body)
+            print("The index %s was successfully created" % (self.index_name))
```
Yes, I modified it so exceptions are handled

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+Arguments:
+    - index_body: The body containing the configurations of the index (used for index creation)
+    - index_name: The index name to create/ingest/delete
+    - client: an OpenSearch Python client object
```
Done, every comment is in tab order

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+        self.index_body = index_body
+        self.client = client
+
+    # Checks if an index exists; if not, the index is created using the OpenSearch Python client (client is defined in the startup/refresh jobs)    
```
Done, every comment is in tab order and uses multiline comment strings

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+        else:
+            print("The index %s exists already" % (self.index_name))
+
+    # Checks if an index exists; if it does, the index is deleted using the OpenSearch Python client (client is defined in the startup/refresh jobs)          
```
Done, modified all parameters with one specific datatype to have a type definition

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+    # Checks if an index exists; if it does, the index is deleted using the OpenSearch Python client (client is defined in the startup/refresh jobs)          
+    def delete_index(self):
+        if self.client.indices.exists(index = self.index_name):
+            self.client.indices.delete(index = self.index_name)
```
Error handling is implemented for each function of `SampleDataIndex`

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+        else:
+            print("The index %s does not exist" % (self.index_name))
+
+    # Calls the ingest() function to ingest more data using provided key word arguments
```
Done, every comment is in tab order and uses multiline comment strings

huyaboo commented on src/playground/sample_data_ingestion/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,34 @@
+
+    # Calls the ingest() function to ingest more data using provided key word arguments
+    def ingest_more(self, **kwargs):
+        ingest(client = self.client, **kwargs)
```
I implemented error handling with the `ingest_more`

huyaboo commented on src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py:
```diff
@@ -0,0 +1,145 @@
+        self.headers = {
+            'Authorization': __authorization,
+            'Content-Type': 'application/json'
+        }
```
Yes, `Authentication` is now an abstract base class. For my example, I've implemented the `BasicAuthentication` class to handle my use case.

huyaboo commented:
@bandinib-amzn I've addressed all comments but I will make a quick note that Python doesn't enforce type checking (duck typing). I kept in my custom input validation to enforce type checking, however, as my unit tests require them.

bandinib-amzn commented on src/playground/sample_data_tooling/sample_data_authentication/sample_data_authentication.py:
```diff
@@ -0,0 +1,25 @@
+        Returns:
+            - Requires a dict to be returned
+        """
+        raise NotImplementedError("get_auth() should be implemented for every Authentication subclass")
```
nit: add new line

huyaboo replied:
Done 👍🏽

bandinib-amzn commented on src/playground/sample_data_tooling/sample_data_authentication/sample_data_basic_authentication.py:
```diff
@@ -0,0 +1,38 @@
+        return {
+            'Authorization': __authorization,
+            'Content-Type': 'application/json'
+        }
```
nit: add new line

huyaboo replied:
Done 👍🏽

bandinib-amzn commented on src/playground/sample_data_tooling/sample_data_commons/validate_filename.py:
```diff
@@ -0,0 +1,20 @@
+    if type(filename) is not str:
+        raise TypeError("filename should be a string")
+    if ".json" not in filename and ".csv" not in filename:
+        raise ValueError("filename must be a .json or .csv file")
```
add new line

huyaboo replied:
Done 👍🏽

bandinib-amzn commented on src/playground/sample_data_tooling/sample_data_indices/sample_data_indices.py:
```diff
@@ -0,0 +1,91 @@
+        except Exception as e:
+            print(e)
+            print("Check configurations again; additionally, the index should be created before it can be ingested")
+            raise
```
nit: fix no newline at end of file

huyaboo replied:
Done 👍🏽

bandinib-amzn reviewed (approved):
Overall looks good to me! Just add new line at the end of each file.
